# Tarea 2 - Paralel Computing 
<b>Autores: </b> Mauricio Ibáñez & César Núñez
<b>Fecha: </b> 10/09/2023


## Parte I
### 1.	¿Qué es un proceso embarrassingly parallel y uno inherentemente serial? Dé un ejemplo de cada uno (distintos a los vistos en clase)

Un proceso embarrasingly parallel (EP) está caracterizado por la facilidad inherente que tiene de poder descomponerse en distintos subprocesos a desarrollarse en paralelo. En contraste, un proceso inherentemente serial (IS) es aquel que no puede paralelizarse para nada. La principal diferencia está entre la dependencia de información entre tareas: por un lado, en los procesos EP se pueden generar subprocesos o tareas que son independientes entre sí y no requieren información entre tareas; mientras que en los procesos IS sí existe una fuerte relación entre una tarea y otra, generando la necesidad de que estas se comuniquen frecuentemente para poder obtener el resultado final.

Un ejemplo de un proceso EP es la decodificación de una imagen hacia una base de datos. Esta se puede tranquilamente particionar en n subgrupos y cada uno de ellos operar de manera independiente, identificando los colores de cada pixel por separado. Estas subtareas no dependen entre sí, por lo que es completamente paralelizable.
Un ejemplo de un proceso IS es una sucesión de números a partir de un dato inicial o también una estimación econométrica. En el caso de esta última, no puedes dividir los datos para obtener la estimación por lo que sólo puede conseguirse el resultado final utilizando toda la información en un solo código 

### 2.	¿Cuáles son los dos principales cuellos de botella al paralelizar un proceso? Explíquelo a partir de las leyes de Amdahl y Gustafson 

Los dos principales cuellos de botella en un proceso son **i) fracción no paralelizable** **ii) secuencialidad**. En el caso de la fracción no paralelizable, esto se puede observar a través de la Ley de Amdahl dado que limitaráel speedup de la tarea a mayor proporción no paralelizable. Respecto a la secuencialidad, de acuerdo la Ley de Gustafson, un proceso se ve ralentizado si la ejeución de tareas se ejecuta serialmente dado que unos resultados dependen de otros anteriormente ejecutados.

### 3.	Describa los recursos (CPU y GPU) de su computadora y provea evidencia (ie. screenshot). 

**CPU:**
- 4 núcleos
- 8 procesadores (2 en cada núcleo)
- Velocidad del procesador: 2.3GHz
GPU:
Procesador: intel core I7 



<img src=https://raw.githubusercontent.com/cesarnunezh/Diplomado_PUCP/main/Big%20Data/Tarea%202/CPU.PNG>



**GPU:**
- Memoria GPU total: 7.9GB
- Memoria compartida: 5.9 GB

### 4.	¿En qué se diferencia un CPU de un GPU? Dé un ejemplo de un proceso que convendría paralelizar en cada uno de estos tipos de unidad de procesamiento. 

In [ ]:
#MAU

### 5.	Se sabe estima que la nueva memoria RAM POWER 3000 permitirá acelerar las operaciones de su computadora en la mitad del tiempo. ¿Qué tanto más rápido será la velocidad de ejecución de un programa si el 77% del mismo se dedica a operaciones inherentemente paralelizables?

Se estima la velocidad de ejecución (_Speedup_) en base a la Ley de Amdahl:

$S_{latency}(s) = \cfrac{1}{(1 - p)+ (\cfrac{p}{s})}$
 
donde $p$ = 0.77, $s$ = 2 dado que la nueva memoria RAM acelera la operación en la mitad del tiempo. Se ejecuta y obtiene:
p = 0.77
s = 2

speed = 1/((1-p)+ (p/s))
speed

In [6]:
p = 0.77
s = 2

speed = 1/((1-p)+ (p/s))
speed

1.6260162601626016

la velocidad de ejecución del programa será aproximadamente 1.626 veces más rápida cuando se utilice la nueva memoria RAM y se paraleliza el 77% de las tareas

### 6.	Piense en una tarea serial que le han encargado paralelizar. Describa el diseño de la implementación en paralelo de dicha tarea siguiente el método de Foster y los cuatro elementos que lo componen.
El Método de Foster consiste en estructurar el problema en cuatro estapas: particionar (dividir las tareas en los procesadores), comunicar (determinar y realizar las comunicaciones necesarias para que siga el código), agregar (agrupar las tareas y comunicaciones en tareas más grandes, priorizando aquellas que están relacionadas) y mapping (asignar las tareas ya agregadas a los procesadores).
Suponiendo que la tarea es el scrappeo de la base de datos de la SBS respecto de los montos de créditos otorgados a personas naturales y jurídicas en los meses comprendidos entre 2008 y 2023 para la posterior elaboración de tablas y gráficos resúmenes, es posible realizar una paralelización siguiendo el método de Foster.

En primer lugar, debemos identificar qué tareas pueden particionarse. Las tareas particionables serían: scrappeo de la web de la SBS para los periodos solicitados y la transformación de cada tabla extraída para una posterior unión, se podría desagregar la data entre el número de los procesadores lógicos disponibles. Además, también podría paralelizarse la elaboración de gráficos o tablas resúmenes una vez que se tenga la base de datos completa. En segundo lugar, se puede identificar que las comunicaciones necesarias entre procesadores es principalmente entre para cuando se requiera unificar las bases de datos y comunicar la base de datos completa para la elaboración de los gráficos y tablas resumen. En tercer lugar, la agrupación de tareas sólo consistiría en el scrappeo, transformación y la unión de los subperiodos que corresponderían a cada procesador. En cuarto lugar, debe asignarse los subperiodos a cada procesador lógico y determinar cuál de los procesadores será el procesador principal que realice la unión de datos final; así como la asignación de tablas y gráficos en cada uno de los procesadores. 

### 7.	¿Cuál es la diferencia entre point-to-point communication y collective communication en MPI? ¿Cuál es la diferencia entre las operaciones de broadcasting, gathering y scattering?
El point-to-point communication es aquella comunicación que va desde un procesador lógico hacia otro específico, pero no pasa por ningún otro procesador; mientras que el collectiv comunication es aquella comunicación que va de un procesador lógico hacia más de un procesador lógico o incluso todos.

El broadcast es un tipo de comunicación colectiva en la cual un procesador lógico envía información hacia el resto de procesadores y todos reciben la misma información. Esto ocurre cuando una información generada en un núcleo es importante para las tareas del resto de núcleos.
El scatter es un tipo de comunicación colectiva en la cual un procesador lógico envía distintas piezas de información al resto de procesadores y cada uno recibe una información diferente. Esto es útil cuando buscamos paralelizar una tarea que depende de una información almacenada en un único procesador.
El gather es un tipo de comunicación colectiva en la cual un procesador lógico recupera la información de parte del resto procesadores lógicos. Esto es útil para consolidar la información de acciones paralelizadas que requieren un procesamiento conjunto de manera posterior.

## Parte II
### 1.	(3 puntos) Escribir un código (“parte2_1.py”) que realice lo siguiente:
#### a.	Que un procesador genere una lista y lo envíe a otros tres procesadores.
#### b.	Que cada uno de los tres procesadores reciba la lista enviada, imprima su número de procesador y la lista. 


In [ ]:
#MAU

#### c.	En otro chunk responda: De ejecutar el código 100 veces, ¿el orden de los resultados será siempre igual? ¿Por qué?

In [ ]:
#MAU

### 2.	(2 puntos) Escribir un código (“parte2_2.py”) que realice lo siguiente:
#### a.	Usando la opción de broadcasting en MPI, defina un diccionario de cuatro elementos que contenga sus cuatro cursos favoritos del diplomado desde el primer procesador. Repita esto para todos los procesadores disponibles. Luego, registre y compare el tiempo de demora desde cada procesador.
#### b.	Defina una secuencia de valores: {0,1,2,3,…, n} en donde n es el número de procesadores de su computadora desde uno de los núcleos. Luego disperse los valores a cada uno de los núcleos restantes usando la opción scattering. Identifique si el número asignado por el proceso coincide con el rango del procesador. 

In [ ]:
#MAU

### 3.	(5 puntos) Generar un numpy array que almacene el archivo: “tarea2.csv” del siguiente link. Luego:

In [1]:
import numpy as np
import os
os.chdir(r"D:\1. Documentos\0. Bases de datos\0. Dofiles y Scripts\Diplomado_PUCP\Big Data\Tarea 2")
narray = np.genfromtxt("tarea2.csv", delimiter= ",", dtype=float)
narray = narray[~np.isnan(narray)]

In [2]:
max = np.max(narray)
print(max)

4448.366507


#### a.	Escribir un código que halle el valor máximo de “tarea2” usando un procesador. Imprimir el valor máximo. Registrar el tiempo de demora.

In [29]:
%%writefile pregunta3a.py
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = MPI.COMM_WORLD.Get_rank()
size = comm.Get_size()

import numpy as np
import os
os.chdir(r"D:\1. Documentos\0. Bases de datos\0. Dofiles y Scripts\Diplomado_PUCP\Big Data\Tarea 2")
narray = np.genfromtxt("tarea2.csv", delimiter= ",", dtype=float)
narray = narray[~np.isnan(narray)]
if rank == 0:
    max_1 = np.max(narray)
    print(max_1)


Overwriting pregunta3a.py


In [30]:
import time
start = time.time()
! mpiexec -n 8 python pregunta3a.py
end = time.time()
tiempo = end - start
print(f'El tiempo promedio de ejecución es de {tiempo}s')

4448.366507
El tiempo promedio de ejecución es de 0.7967944145202637s


#### b.	Escribir un código que realice las siguientes indicaciones : (i) dividir el numpy en dos partes iguales, (ii) que dos procesadores distintos encuentren el máximo de cada parte, y (iii) que otro procesador junte los máximos hallados y encuentre el máximo global. Registrar el tiempo de demora.

In [125]:
%%writefile pregunta3b.py
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = MPI.COMM_WORLD.Get_rank()
size = comm.Get_size()

import numpy as np
import os


if rank == 0:
    os.chdir(r"D:\1. Documentos\0. Bases de datos\0. Dofiles y Scripts\Diplomado_PUCP\Big Data\Tarea 2")
    narray = np.genfromtxt("tarea2.csv", delimiter= ",", dtype=float)
    narray = narray[~np.isnan(narray)] 
    len_narray = int(len(narray)/3)
else:
    narray = None
    len_narray = None

len_narray = comm.bcast(len_narray, root=0)

local_data = np.empty(len_narray, dtype=float)
comm.Scatter(narray, local_data, root=0)
    
print(rank)
print(local_data)
print(f'{len_narray}')


'''
local_data = np.empty(len_narray, dtype=float)
comm.Scatter(narray, local_data, root=0)
local_max = np.max(local_data)

max_total = comm.gather(local_max, root=0)

if rank == 0:
    max_total = max(max_total)
    print(max_total)
'''

Overwriting pregunta3b.py


In [126]:
import time
start = time.time()
! mpiexec -n 3 python pregunta3b.py
end = time.time()
tiempo = end - start

print(f'El tiempo promedio de ejecución es de {tiempo}s')

^C
El tiempo promedio de ejecución es de 3.4522597789764404s


#### c.	Repetir 3b dividiendo el numpy original en tres partes. Registrar el tiempo de demora.

In [7]:
%%writefile pregunta3c.py
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = MPI.COMM_WORLD.Get_rank()
size = comm.Get_size()

import numpy as np
import os
os.chdir(r"D:\1. Documentos\0. Bases de datos\0. Dofiles y Scripts\Diplomado_PUCP\Big Data\Tarea 2")
narray = np.genfromtxt("tarea2.csv", delimiter= ",", dtype=float)
narray = narray[~np.isnan(narray)]

if rank == 0:
    length = int(len(narray)/3)
    max_1 = np.max(narray[:length])
    comm.send(max_1, dest=3 , tag=3)
elif rank ==1:
    length_1 = int(len(narray)/3)
    length_2 = int(2*len(narray)/3)
    max_2 = np.max(narray[length_1:length_2])
    comm.send(max_2, dest=3 , tag=3)
elif rank ==2:
    length_2 = int(2*len(narray)/3)
    max_3 = np.max(narray[length_2:])
    comm.send(max_3, dest=3 , tag=3)
elif rank ==3:
    max_1 = comm.recv(source=0 , tag=3)
    max_2 = comm.recv(source=1 , tag=3)
    max_3 = comm.recv(source=2, tag=3)
    max_total = max(max_1,max_2,max_3)
    print(max_total)

Overwriting pregunta3c.py


In [8]:
import time
start = time.time()
! mpiexec -n 8 python pregunta3c.py
end = time.time()
tiempo = end - start

print(f'El tiempo promedio de ejecución es de {tiempo}s')

4448.366507
El tiempo promedio de ejecución es de 0.9440469741821289s


#### d.	Comparar los tiempos registrados en 3a, 3b y 3c. ¿Hay una reducción del tiempo? ¿La reducción del tiempo es lineal? ¿Por qué?
En la pregunta a) el tiempo obtenido es de 0.7334s, en la pregunta b) es de 0.7272s y en la pregunta c) es de 0.7094s. Aunque, a veces, cuando he corrido nuevamente los datos salen distintas duraciones.